### PR TITLE
Cache access token on disk

### PR DESCRIPTION
**Pull request by NRHelmi**
This PR adds support to cache `oauth` access token on disk

### CODE REVIEW

**Comment by NRHelmi** [comment]
> I lack context for this PR, and I'm a bit swamped, so i don't feel able to review this. can someone else please pick this up?
> 
> Also, @NRHelmi , the tests are currently failing

thanks @NHDaly checking what's wrong

**Comment by NHDaly** [comment]
What's the word on this PR? Is someone else going to review it? @NRHelmi can you track down a reviewer?

File: src/rest.jl [hunk 1]
```diff
@@ -81,7 +81,51 @@ function get_access_token(ctx::Context, creds::ClientCredentials)::AccessToken
...
+
+# cache name
+function _cache_file()
+    return joinpath(homedir(), ".rai", "tokens.json")
+end
+
+# read oauth cache
+function _read_cache()
+    try
+        return copy(JSON3.read(read(_cache_file(), String)))
```
Comment (quinnj) [inline src/rest.jl:95]:
```suggestion
        return copy(JSON3.read(read(_cache_file())))
```
JSON3.read can accept bytes directly; no need for an extra copy.

**Comment by quinnj** [comment]
This looks generally good; can I request a change in style though? I think the pattern of:
```julia
try
   # do stuff
catch
end
```
, while convenient, can lead to some frustrating experiences later on. If there happens to be a typo, or non-100% test coverage, or something else changes in the future (like the exact JSON structure), then it becomes _really_ hard to "be notified" that there's something wrong, since we're just swallowing errors.

I'd prefer to be more explicit; let's check `isfile` directly on the `tokens.json`. It's fine to return `nothing` in the error case, but then let's check that specifically instead of trying to call getindex like `creds[:token]`.

Moving to this style may mean dealing with more error cases in the short-term, but as we add additional error handling, it will make our code more explicit and predictable long-term.

**Comment by NRHelmi** [comment]
thanks @NHDaly for the checkout, yep I believe @geokollias is planning on reviewing this PR but he probably didn't have enough bandwidth to check.
thanks @quinnj for the comments will address your request :pray:

**Comment by NRHelmi** [comment]
Hi @quinnj I addressed your code change request, can you please check the PR again if you have some bandwidth :pray:

File: src/rest.jl [hunk 2]
```diff
@@ -81,7 +81,64 @@ function get_access_token(ctx::Context, creds::ClientCredentials)::AccessToken
...
+    end
+end
+
+# Read access token from cache
+function _read_token_cache(creds::ClientCredentials)
+    try
+        cache = _read_cache()
+        cache === nothing && return nothing
+
+        if creds.client_id in keys(cache)
```
Comment (quinnj) [inline src/rest.jl:112]:
```suggestion
        if haskey(cache, creds.client_id)
```
should we also add an expiration check here and return `nothing` if the cached token is expired? I can imagine the workflow where you infrequently run things, this token may end up being stale, so it seems better to force the request of a new one.

But maybe we already do that check later, so it doesn't matter here?

**Review by quinnj — APPROVED** [review]
Yeah, this is looking good.

One request: can we add a simple unit test for the new `_write_token_cache` and `_read_token_cache` functions? We're still wrapping those method bodies in uncondtional `try-catch`, so it'd be nice to at least have simple unit tests that assert that we're writing out what we expect and that we can read it back in again appropriately. I know they're internal-ish functions, but having the unit tests are still really helpful for the reason I mentioned above and they're easy to remove/change in the future if we refactor things.

File: src/rest.jl (hunk 2 above)
Comment (NRHelmi) [reply src/rest.jl:112]: you are right, the workflow is the following: we get the cache -> check if token expired (fetch new one and save it to cache if expired) -> proceed

**Comment by NRHelmi** [comment]
> Yeah, this is looking good.
> 
> One request: can we add a simple unit test for the new `_write_token_cache` and `_read_token_cache` functions? We're still wrapping those method bodies in uncondtional `try-catch`, so it'd be nice to at least have simple unit tests that assert that we're writing out what we expect and that we can read it back in again appropriately. I know they're internal-ish functions, but having the unit tests are still really helpful for the reason I mentioned above and they're easy to remove/change in the future if we refactor things.

thanks @quinnj will do :pray:

**Pull request merged**